migration/migration_blackbox_test.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/almighty/almighty-core/resource"
+
+	"github.com/jinzhu/gorm"
+	_ "github.com/lib/pq"
+	errs "github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+// fn defines the type of function that can be part of a migration steps
+type fn func(tx *sql.Tx) error
+
+const (
+	// Migration version where to start testing
+	initialMigratedVersion = 45
+	databaseName           = "test"
+)
+
+var (
+	conf       *config.ConfigurationData
+	migrations migration.Migrations
+)
+
+func setupTest() {
+	var err error
+	conf, err = config.GetConfigurationData()
+	if err != nil {
+		panic(fmt.Errorf("Failed to setup the configuration: %s", err.Error()))
+	}
+	configurationString := fmt.Sprintf("host=%s port=%d user=%s password=%s sslmode=%s connect_timeout=%d",
+		conf.GetPostgresHost(),
+		conf.GetPostgresPort(),
+		conf.GetPostgresUser(),
+		conf.GetPostgresPassword(),
+		conf.GetPostgresSSLMode(),
+		conf.GetPostgresConnectionTimeout(),
+	)
+
+	db, err := sql.Open("postgres", configurationString)
+	defer db.Close()
+	if err != nil {
+		panic(fmt.Errorf("Cannot connect to database: %s\n", err))
+	}
+
+	db.Exec("DROP DATABASE " + databaseName)
+
+	_, err = db.Exec("CREATE DATABASE " + databaseName)
+	if err != nil {
+		panic(err)
+	}
+
+	migrations = migration.GetMigrations()
+}
+
+func TestMigrations(t *testing.T) {
+	resource.Require(t, resource.Database)
+
+	setupTest()
+
+	configurationString := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=%s connect_timeout=%d",
+		conf.GetPostgresHost(),
+		conf.GetPostgresPort(),
+		conf.GetPostgresUser(),
+		conf.GetPostgresPassword(),
+		databaseName,
+		conf.GetPostgresSSLMode(),
+		conf.GetPostgresConnectionTimeout(),
+	)
+
+	db, err := sql.Open("postgres", configurationString)
+	defer db.Close()
+	if err != nil {
+		panic(fmt.Errorf("Cannot connect to DB: %s\n", err))
+	}
+	gormDB, err := gorm.Open("postgres", configurationString)
+	defer gormDB.Close()
+	if err != nil {
+		t.Fatalf("Cannot connect to DB: %s\n", err)
+	}
+	dialect := gormDB.Dialect()
+	dialect.SetDB(db)
+
+	m := migrations[:initialMigratedVersion]
+
+	for nextVersion := int64(0); nextVersion < int64(len(m)) && err == nil; nextVersion++ {
+		var tx *sql.Tx
+		tx, err = db.Begin()
+		if err != nil {
+			t.Fatalf("Failed to start transaction: %s\n", err)
+		}
+
+		if err = migration.MigrateToNextVersion(tx, &nextVersion, m, databaseName); err != nil {
+			t.Errorf("Failed to migrate to version %d: %s\n", nextVersion, err)
+
+			if err = tx.Rollback(); err != nil {
+				t.Fatalf("error while rolling back transaction: ", err)
+			}
+			t.Fatal("Failed to migrate to version after rolling back")
+		}
+
+		if err = tx.Commit(); err != nil {
+			t.Fatalf("Error during transaction commit: %s\n", err)
+		}
+	}
+
+	// Insert dummy test data
+	assert.Nil(t, runSQLscript(db, "044-insert-test-data.sql"))
+
+	// Migration 45
+	migrationToVersion(db, migrations[:(initialMigratedVersion+1)], (initialMigratedVersion + 1))
```

Comment:
@aslakknutsen I did these 4 migrations to draft some examples of the idea. Nothing else than that.

I agree that the value of column assertions of indexes might not be too hight. But you can also tests that the migration did what it was supposed to do like checking the defaults values for the new column were defined, etc... The column assertion was an example...